### PR TITLE
[core] TD validation in produce() method while exposing things

### PR DESCRIPTION
**Pull request by fillobotto**
Now `WoT.produce()` validates the TD according [this specification](https://w3c.github.io/wot-scripting-api/#validating-an-exposedthinginit). Related test was also added for completeness.

### CODE REVIEW

File: packages/core/src/wot-impl.ts [hunk 1]
```diff
@@ -86,7 +109,10 @@ export default class WoTImpl {
 
                 // FIXME should be constrained version that omits instance-specific parts (but keeps "id")
                 let template = td;
```
Comment (relu91) [inline packages/core/src/wot-impl.ts:90]: could you please deepclone the TD object? I think it is best practice not to mess around the inputs

File: packages/core/src/wot-impl.ts [hunk 2]
```diff
@@ -15,14 +15,14 @@
...
 import * as TD from "@node-wot/td-tools";
-
 import Servient from "./servient";
 import ExposedThing from "./exposed-thing";
 import ConsumedThing from "./consumed-thing";
-import Helpers from "./helpers";
-import { ContentSerdes } from "./content-serdes";
+import Ajv from 'ajv';
+
+const ajv = new Ajv({strict:false, allErrors: true});
```
Comment (relu91) [inline packages/core/src/wot-impl.ts]: `allErrors` should be removed

File: packages/core/src/wot-impl.ts [hunk 3]
```diff
@@ -74,6 +74,29 @@ export default class WoTImpl {
         }
     }
 
+    validateThingDescription(td: any) {
```
Comment (relu91) [inline packages/core/src/wot-impl.ts]: Rename to, `createExposeThingInitSchema`

File: packages/core/src/wot-impl.ts [hunk 4]
```diff
@@ -86,7 +109,19 @@ export default class WoTImpl {
...
+                let tdSchema = TDSchema;
+                let exposeThingInitSchema = this.validateThingDescription(tdSchema);
+
+                const validate = ajv.compile(exposeThingInitSchema)
+
+                if(!validate(template)) {
+                    throw new Error("Thing Description JSON schema validation failed: " + JSON.stringify(validate.errors));
+                }
+
+                this.validateThingDescription(template);
```
Comment (relu91) [inline packages/core/src/wot-impl.ts]: to be removed

File: packages/core/src/wot-impl.ts [hunk 5]
```diff
@@ -86,7 +109,19 @@ export default class WoTImpl {
 
                 // FIXME should be constrained version that omits instance-specific parts (but keeps "id")
                 let template = td;
+
+                let tdSchema = TDSchema;
+                let exposeThingInitSchema = this.validateThingDescription(tdSchema);
+
+                const validate = ajv.compile(exposeThingInitSchema)
```
Comment (relu91) [inline packages/core/src/wot-impl.ts]: Move everything  at the module root level

File: packages/core/src/wot-impl.ts [hunk 6]
```diff
@@ -86,7 +109,19 @@ export default class WoTImpl {
...
+
+                if(!validate(template)) {
+                    throw new Error("Thing Description JSON schema validation failed: " + JSON.stringify(validate.errors));
```
Comment (relu91) [inline packages/core/src/wot-impl.ts]: Can we pretty-print the errors? Somehow?

File: packages/core/src/wot-impl.ts [hunk 7]
```diff
@@ -74,6 +74,29 @@ export default class WoTImpl {
         }
     }
 
+    validateThingDescription(td: any) {
+        if(td.required !== undefined) {
+            let reservedKeywords: Array<string> = [ 
+                "title", "@context", "instance", "forms", "security", "href"
```
Comment (relu91) [inline packages/core/src/wot-impl.ts]: add `securityDefinitions`

File: packages/core/test/ServerTest.ts [hunk 8]
```diff
@@ -90,6 +95,33 @@ class WoTServerTest {
         expect(thing).to.have.property("properties").to.have.property("myProp");
     }
 
+    @test async "validate thing description"() {
```
Comment (relu91) [inline packages/core/test/ServerTest.ts]: to be removed, or better to add a specific test for validation (e.g. in the helpers class)

**Review by relu91 — CHANGES_REQUESTED** [review]

File: packages/core/src/helpers.ts [hunk 9]
```diff
@@ -173,4 +178,56 @@ export default class Helpers {
...
+    const td = JSON.parse(JSON.stringify(tdSchema));
+
+    if(data["@type"] !== undefined && data["@type"] == "tm:ThingModel") {
+      return {
+        valid: false,
+        errors: "ThingModel declaration is not supported"
+      };
+    }
+
+    let exposeThingInitSchema = Helpers.validateThingDescription(td);
```
Comment (relu91) [inline packages/core/src/helpers.ts]: One thing that I don't still like, is the fact that each time we perform the `createExposeThingInitSchema` algorithm (i.e. validateThingDescription in your code). can we move :
 ```ts
let exposeThingInitSchema = Helpers.validateThingDescription(td);
const validate = ajv.compile(exposeThingInitSchema);
```
to a dedicated const outside of the body of `validateExposedThingInit`?

**Review by relu91 — CHANGES_REQUESTED** [review]
Keep in mind also the `deepClone` issue

File: packages/core/src/helpers.ts [hunk 10]
```diff
@@ -209,7 +209,16 @@ export default class Helpers {
    * Helper function to validate an ExposedThingInit
    */
   public static validateExposedThingInit(data : any) {
-    let exposeThingInitSchema = Helpers.validateThingDescription(JSON.parse(JSON.stringify(tdSchema)));
+    const td = JSON.parse(JSON.stringify(tdSchema));
+
+    if(data["@type"] !== undefined && data["@type"] == "tm:ThingModel") {
```
Comment (relu91) [inline packages/core/src/helpers.ts]: `undefined === "onestring"` is always false. So you can skip the first check

File: packages/core/src/helpers.ts (hunk 9 above)
Comment (fillobotto) [reply packages/core/src/helpers.ts]: Please see new changes. Honestly I can't come with a cleaner solution. it's probably due to my lackings in TS :/
Let me know if you have better suggestions to the problem

File: packages/core/src/helpers.ts [hunk 11]
```diff
@@ -173,4 +182,51 @@ export default class Helpers {
     }
     return value;
   }
+
+  /**
+   * Helper function to remove reserved keywords in required property of TD JSON Schema
+   */
+  static validateThingDescription(td: any) {
```
Comment (relu91) [inline packages/core/src/helpers.ts]: Can you rename this to `createExposeThingInitSchema(tdSchema: unknown)`

**Review by relu91 — CHANGES_REQUESTED** [review]
Well done! minor detail and we are almost there. + I would like to add that deep cloning we discussed about.

File: packages/core/src/wot-impl.ts (hunk 1 above)
Comment (relu91) [reply packages/core/src/wot-impl.ts:90]: @fillobotto check this.

**Comment by relu91** [comment]
To complete the validation we have to remember that we should check if `tm:ThingModel` is present or/and `tm:ref` and `tm:extends`

**Review by relu91 — CHANGES_REQUESTED** [review]
We are still missing the deep clone functionality in the produce method. I think it makes sense to bundle the change to this PR cause it already refactors the produce method.

**Review by relu91 — APPROVED** [review]
Never mind the TD is cloned inside the ExposedThing constructor.

**Pull request merged**